[Mime] Add Address::fromString

gisostallenberg opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #33086
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/12128

This will allow to create a Address from a string such as 'Name <name@example.com>'
Example: 
```php
$address = Address::fromString("Name <name@example.com>");
```

Shkarsardar commented:
can you explain it? how the feature should work

gisostallenberg commented:
@Shkarsardar see https://github.com/symfony/symfony-docs/pull/12128

fabpot commented on src/Symfony/Component/Mime/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Added PHPUnit constraints
  * Added `AbstractPart::asDebugString()`
+ * Added NamedAddress::fromString
```
Can you quote it with ` and add `()` at the end (like the line just above)

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+            static::FROM_STRING_PATTERN,
+            $string,
+            $matches
+        );
```
Should be on one line and could be moved to the condition: `if (!$matched = $matched = preg_match(static::FROM_STRING_PATTERN, $string, $matches) {`

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+        return new static(
+            $matches['address'],
+            trim($matches['name'], ' \'"')
+        );
```
should be on one line.

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+            throw new \RuntimeException(sprintf('Could not parse (%s) to a named address', $string));
+        }
+
+        return new static(
```
I suppose that we expect a encoded address, so we must decode the parts here, right?

gisostallenberg replied:
You mean the inverse of `Rfc2231Encoder`?

fabpot replied:
I was more thinking about IDN. Not sure if this is needed though.

gisostallenberg replied:
As `NamedAddressTest::testConstructor` shows you can also provide an un-encoded address.

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -16,6 +16,8 @@
  */
 final class NamedAddress extends Address
 {
+    const FROM_STRING_PATTERN = '~(?<name>.*)<(?<address>[^>]+)>~';
```
I think the RFC for an address is more complex than this one. We should at least support the most common forms.

gisostallenberg replied:
You mean the inverse of `Rfc2231Encoder`?

derrabus commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+    public static function fromString(string $string): self
+    {
+        if (false === strpos($string, '<')) {
+            return new static($string, $string);
```
```suggestion
            return new self($string, $string);
```

The class is final anyway. 😃

derrabus commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+            throw new \RuntimeException(sprintf('Could not parse (%s) to a named address', $string));
+        }
+
+        return new static(
```
```suggestion
        return new self(
```

Same here.

derrabus commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +41,25 @@ public function toString(): string
+        }
+
+        $matched = preg_match(
+            static::FROM_STRING_PATTERN,
```
```suggestion
            self::FROM_STRING_PATTERN,
```

derrabus commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -16,6 +16,8 @@
  */
 final class NamedAddress extends Address
 {
+    const FROM_STRING_PATTERN = '~(?<name>.*)<(?<address>[^>]+)>~';
```
```suggestion
    private const FROM_STRING_PATTERN = '~(?<name>.*)<(?<address>[^>]+)>~';
```

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -16,6 +16,8 @@
  */
 final class NamedAddress extends Address
 {
+    private const FROM_STRING_PATTERN = '~(?<name>.*)<(?<address>[^>]+)>~';
```
We should have a look at RFC 2822 to be sure we cover most common use cases.

gisostallenberg replied:
I think I'm going to need some help here. As far as I understand from https://tools.ietf.org/html/rfc2822#section-3.4 the `angle-addr` is simply matched with this regex, but `addr-spec` might contain a < which is not checked in this regex (as in 'Me <"name-with<address"@domain.com>) which is not 'common' I think. Any more checking should be done while construction the (Named)Address, which is already done.

sstok replied:
You are a brave soul. Welcome to email addresses 😄 funny enough you can use any character in the email address (and name part) as long as it's quoted, and better yet you can even use quotes in quotes, as long as those are properly escaped! I believe you can even use `@` in the name part 😐

Parsing a mailbox header with a regex is like parsing HTML with a regex, possible but you are gonna miss a lot edge-cases.

During the Fossa hackaton while struggling to come-up with a way to parse an email message (and its headers) I came discover that you can only properly parse a MIME header using a full fledged lexer, as there are so many edge cases. _But, luckily I managed to solve my use-case another way 😊 -- that didn't involve a complete message parser._

If we don't want to support all the most complex cases, it's properly best to look for the first and last `<>` and assume this as the actual address, and use everything before the first `<` as the name. That is safest and fastest.

fabpot replied:
I haven't checked, but can't we have comments as well? We can probably don't implement them though. If we keep the simple regex, let's at least add a comment to describe what we are doing and what we don't want to support.

sstok replied:
Ah yes comments 😭 

> If we keep the simple regex, let's at least add a comment to describe what we are doing and what we don't want to support.

👍 

Plus. Anyone who must use a complex name plus address can still construct the object manually.

gisostallenberg replied:
> it's properly best to look for the first and last <> and assume this as the actual address, and use everything before the first < as the name. That is safest and fastest.

I'll update the regex to match the first `<` to the last `>`.


> If we keep the simple regex, let's at least add a comment to describe what we are doing and what we don't want to support.

I'll add that comment

gisostallenberg replied:
Thanks @sstok for helping me. I updated the regex and added a comment

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +48,17 @@ public function toString(): string
+            return new self($string, '');
+        }
+
+        if (!(preg_match(self::FROM_STRING_PATTERN, $string, $matches))) {
```
extra parenthesis could be removed.

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +48,17 @@ public function toString(): string
+        }
+
+        if (!(preg_match(self::FROM_STRING_PATTERN, $string, $matches))) {
+            throw new \RuntimeException(sprintf('Could not parse (%s) to a named address', $string));
```
Acccording to our CS: `Could not parse "%s" to a "NamedAddress" instance.` (notice the dot at the end).

gisostallenberg commented:
Thanks so much for all the effort you put in checking this.

OskarStark commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +48,17 @@ public function toString(): string
+        }
+
+        if (!preg_match(self::FROM_STRING_PATTERN, $string, $matches)) {
+            throw new \RuntimeException(sprintf('Could not parse "%s" to a "NamedAddress" instance.', $string));
```
I propose to use an InvalidArgumentException here.
Not sure it should be in the Mime namespace.

gisostallenberg replied:
There's one in Mime already, so changed it to that

OskarStark commented on src/Symfony/Component/Mime/Tests/NamedAddressTest.php:
```diff
@@ -24,4 +24,80 @@ public function testConstructor()
     }
+
+    /**
+     * @dataProvider fromStrings
```
fromStringProvider ?

OskarStark commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -11,11 +11,22 @@
+     * config, and allows to have more readable config.
+     * This does not try to cover all edge cases for address.
+     */
+    private const FROM_STRING_PATTERN = '~(?<displayName>[^<]*)<(?<addrSpec>.*)>[^>]*~';
```
Not sure we need this pattern as a constant here, it is only used once

fabpot replied:
The constant allows to have the comment and the regex together, so I think we can keep it like this.

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +50,17 @@ public function toString(): string
+    public static function fromString(string $string): self
+    {
+        if (false === strpos($string, '<')) {
+            return new self($string, '');
```
What about returning a simple `Address` instance here instead?

OskarStark replied:
I am against this, because in this case `fromString()` would have 2 return types NamedAddress|Address 👎

fabpot replied:
But returning a `NamedAddress` is a lie, as there is no name. And the return type should indeed be changed, but to `Address` only (as `NamedAddress` is a subclass of `Address`). What we care about is having an `Address` in your code (the fact that it can be a `NamedAddress` is an implementation details).

Another possibility would be to remove the distinction between `Address` and `NamedAddress` and only have `Address` which would support named addresses.

gisostallenberg replied:
I also did not like having this method returning two kinds. It could also result in an `InvalidArgumentException` as far as I'm concerned.

fabpot replied:
I think throwing an exception is not an option as the use case is configuration, where you don't really know what the user is going to use.

gisostallenberg replied:
But... 
```php
$address = new NamedAddress('example@example.com', '');
```
is allowed at the moment, maybe that should also not be the case?

sstok replied:
> Another possibility would be to remove the distinction between Address and NamedAddress and only have Address which would support named addresses.

👍

gisostallenberg replied:
> > Another possibility would be to remove the distinction between Address and NamedAddress and only have Address which would support named addresses.

Would that be a think to do in this PR, or create a separate PR for?

OskarStark replied:
> What we care about is having an Address in your code (the fact that it can be a NamedAddress is an implementation details).

This is true, but named constructors are not meant to return anything else than the static class itself as AFAIK. It would be strange to me:
```php
public function myMethod(NamedAddress $foo) {
    // ...
}
```
and get an exception when calling it like this, because its not a NamedAdress:
```php
$this->myMethod(NamedAddress::fromString(.....))
```

fabpot replied:
I have done the PR in #33270 as I wanted to understand why I added `NamedAddress` in the first place and the consequences of its removal. It's safe, so let's do it! when #33270 is merged, we will be able to rebase here and the discussion would be close :)

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -39,4 +50,17 @@ public function toString(): string
+        }
+
+        if (!preg_match(self::FROM_STRING_PATTERN, $string, $matches)) {
+            throw new InvalidArgumentException(sprintf('Could not parse "%s" to a "NamedAddress" instance.', $string));
```
Could be `sprintf('Could not parse "%s" to a "%s instance.', $string, static::class)`.

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -11,11 +11,22 @@
+    /**
+     * Simple regex to match a structure like 'Name <email@address.com>'.
+     * It matches anything between the first < and last > as email address.
+     * This will allow to use a single string to construct a NamedAddress, which can be convenient to use in
```
`This allows to use...`?

fabpot commented on src/Symfony/Component/Mime/NamedAddress.php:
```diff
@@ -11,11 +11,22 @@
 final class NamedAddress extends Address
 {
+    /**
+     * Simple regex to match a structure like 'Name <email@address.com>'.
```
`A regex that matches a structure like ...`

fabpot commented:
@gisostallenberg Can you rebase as #33270 has been merged now?

gisostallenberg committed:
Add Address::fromString

This will allow to create an Address from a string such as 'Name <name@example.com>'

fabpot commented:
Thank you @gisostallenberg.
